asf_welsh_energy_consultation/pipeline/process_data.py

```diff
@@ -0,0 +1,256 @@
+
+    Args:
+        variable (str): Variable to split by.
+        new_var_name (str): Name of variable in processed dataset.
+        data (pd.DataFrame, optional): Base data. Defaults to mcs.
+
+    Returns:
+        pd.DataFrame: Cumulative totals dataset.
+    """
+
+    # calculate total number of installations for each date/category pair
+    totals = data.groupby(["commission_date", variable]).size()
+
+    totals = totals.reset_index().rename(columns={0: "sum"})
+
+    idx = pd.date_range(
+        totals["commission_date"].min(), totals["commission_date"].max()
+    )
+
+    totals = totals.pivot(index="commission_date", columns=variable).fillna(0)
+
+    totals.index = pd.DatetimeIndex(totals.index)
+
+    # reindex, fill in gaps with 0, and take cumulative sums
+    cumsums = totals.reindex(idx, fill_value=0).cumsum()
+    cumsums.columns = cumsums.columns.droplevel(0)
+    cumsums = cumsums.rename_axis(None, axis=1).reset_index(names="date")
+
+    cumsums = pd.melt(
+        cumsums,
+        id_vars="date",
+        value_vars=[col for col in cumsums.columns if col != "date"],
+    ).rename(columns={"variable": new_var_name, "value": "Number of heat pumps"})
+
+    cumsums["Number of heat pumps"] = cumsums["Number of heat pumps"].astype(int)
+
+    # remove pre-2015 entries as numbers are small
+    cumsums = cumsums.loc[cumsums.date >= "2015-01-01"].reset_index(drop=True)
+
+    return cumsums
+
+
+# PROCESSING EPC
+
+wales_epc = get_wales_epc()
+
+
+def get_wales_epc_new():
+    """Get first EPC certificates for any property labelled as "new dwelling".
+
+    Returns:
+        pd.DataFrame: New build EPC certificates.
+    """
+    wales_epc_new = (
+        wales_epc.loc[wales_epc["TRANSACTION_TYPE"] == "new dwelling"][
+            ["UPRN", "INSPECTION_DATE", "HP_INSTALLED"]
+        ]
+        .dropna(subset="INSPECTION_DATE")
+        .sort_values("INSPECTION_DATE")
+        .groupby("UPRN")
+        .head(1)
+        .reset_index(drop=True)
+    )
+
+    wales_epc_new["year"] = wales_epc_new["INSPECTION_DATE"].dt.year
+
+    return wales_epc_new
+
+
+def new_hp_counts():
+    """Get counts of new builds with HPs for each year.
+
+    Returns:
+        pd.DataFrame: New build HP counts.
+    """
+    wales_epc_new = get_wales_epc_new()
+    # 2023 not yet complete so drop any post-2022 data
+    wales_epc_new = wales_epc_new.loc[wales_epc_new["INSPECTION_DATE"] < "2023-01-01"]
+
+    new_hp_counts = (
+        wales_epc_new.groupby(["year", "HP_INSTALLED"])
+        .size()
+        .reset_index()
+        .pivot(index="year", columns="HP_INSTALLED")
+    )
+
+    new_hp_counts.columns = new_hp_counts.columns.droplevel(0)
+    new_hp_counts.columns.name = None
+    new_hp_counts = new_hp_counts.rename(columns={False: "Other", True: "Heat pump"})
+
+    new_hp_counts["n"] = new_hp_counts.sum(axis=1)
+    new_hp_counts["prop_hp"] = new_hp_counts["Heat pump"] / new_hp_counts["n"]
+
+    new_builds = new_hp_counts.reset_index()
```

Comment:
Oh I think I changed it here to avoid returning something with the same name as the function - I should probably keep it consistent and just rename the function though